src/box/box.cc

```diff
@@ -1805,6 +1811,230 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+		.vclock_syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.count = vclock_count,
+		.is_timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_vclock_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.vclock_syncs[replica->id] != 0)
+			continue;
+		if (relay_next_vclock_sync(replica->relay,
+					   &data.vclock_syncs[replica->id],
+					   deadline) != 0) {
+			/* Timed out. */
+			trigger_clear(&on_ack);
+			trigger_clear(&on_relay_thread_start);
+			return -1;
+		}
+	}
+
+	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
+	       !data.is_timed_out && !fiber_is_cancelled()) {
+		if (fiber_yield_deadline(deadline))
+			break;
+	}
+
+	trigger_clear(&on_ack);
+	trigger_clear(&on_relay_thread_start);
+	if (fiber_is_cancelled()) {
+		diag_set(FiberIsCancelled);
+		return -1;
+	}
+	if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
+		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
+			 "timed out waiting for remote node vclocks");
+		return -1;
+	}
+	return 0;
+}
+
+/** box_wait_vclock trigger data. */
+struct box_wait_vclock_data {
+	/** Whether the request is finished. */
+	bool is_ready;
+	/** The vclock to wait for. */
+	const struct vclock *vclock;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+};
+
+static int
+box_wait_vclock_f(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct box_wait_vclock_data *data =
+		(struct box_wait_vclock_data *)trigger->data;
+	if (vclock_compare_ignore0(data->vclock, &replicaset.vclock) <= 0) {
+		data->is_ready = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/**
+ * Wait until this instance's vclock reaches @a vclock or @a deadline is
+ * reached.
+ */
+static int
+box_wait_vclock(const struct vclock *vclock, double deadline)
+{
+	if (vclock_compare_ignore0(vclock, &replicaset.vclock) <= 0)
+		return 0;
+	struct trigger on_wal_write;
+	struct box_wait_vclock_data data = {
+		.is_ready = false,
+		.vclock = vclock,
+		.waiter = fiber(),
+	};
+	trigger_create(&on_wal_write, box_wait_vclock_f, &data, NULL);
+	trigger_add(&wal_on_write, &on_wal_write);
+	do {
+		if (fiber_yield_deadline(deadline))
+			break;
+	} while (!data.is_ready && !fiber_is_cancelled());
+	trigger_clear(&on_wal_write);
+	if (fiber_is_cancelled()) {
+		diag_set(FiberIsCancelled);
+		return -1;
+	}
+	if (!data.is_ready) {
+		diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
+			 "timed out");
```

Comment:
Please use `TimedOut` error for timeouts. Here's why:
1. We never use custom errors for timeouts in other places in the code. Inconsistency is never good.
2. Mixing `TimedOut` with `ClientError{FOO_TIMEOUT}` doesn't give the user any new information, actually. Suppose you failed to sync vclock, because socket lags somewhere inside the replication code internals. Technically it's a socket timeout, but you'll return `ClientError{FOO_TIMEOUT}`, which is confusing. Suppose you failed to wait for vclock, because the code syncing vclocks took too much time. Then you'll get `ER_QUORUM_WAIT` although it could just as well be `ER_VCLOCK_WAIT` error.
3. The user/admin may know nothing about quorum/vclock/etc; there's a CALL that failed because the timeout is too low. It could fail with a timeout in some other place inside the CALL. What the user/admin wants to see is `TimedOut` error. What the developer wants to see is file/line/trace where the `TimedOut` error occurred.

Confirmed by @alyapunov.

Please fix the code below to use generic errors, as well:

https://github.com/tarantool/tarantool/blob/e6abe1c91ae36127bedcd3b9a5d5d4705b802c91/src/box/box.cc#L1794-L1804

Reply:
I'd argue about that:
having a "timed out waiting for 5 remote vclocks" error gives the user some additional info. For example, the user might have a wrong quorum setting, making the read poll for all 5 nodes in the cluster. Even when one of them is dead.

Simply having "timed out" error gives no info.

Anyway, I've changed these places:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 56aafafe9..315bed53b 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1944,8 +1944,7 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
                return -1;
        }
        if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
-               diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
-                        "timed out waiting for remote node vclocks");
+               diag_set(TimedOut);
                return -1;
        }
        return 0;
@@ -2001,8 +2000,7 @@ box_wait_vclock(const struct vclock *vclock, double deadline)
                return -1;
        }
        if (!data.is_ready) {
-               diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
-                        "timed out");
+               diag_set(TimedOut);
                return -1;
        }
        return 0;

```

And done the same for box_wait_quorum in a separate commit.

Reply:
> having a "timed out waiting for 5 remote vclocks" error gives the user some additional info. For example, the user might have a wrong quorum setting, making the read poll for all 5 nodes in the cluster. Even when one of them is dead.

In this case, the transaction should fail before the timeout occurs and raise a proper error.

Reply:
I'd say we should try to connect to everyone for a timeout anyway.

Maybe enough connections will be established by that time

Reply:
In this case, you could print a message to the log warning that the configuration might be invalid.

Reply:
I am not in the context since quite a lot has changed in the patch. But if you need any additional info in an error object, you can use `error_append_msg()` and/or `error_set_...()` functions.